{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Evolutionary Similarity Index\n",
    "\n",
    "Set of functions to assess evolutionary similarity between gene families using $I_{ES}$.\n",
    "\n",
    "Accepted input files are: aminoacid [FASTA](https://en.wikipedia.org/wiki/FASTA_format), \n",
    "                                    [NEWICK](https://en.wikipedia.org/wiki/Newick_format), and \n",
    "                                    \".mldist\" from [IQTree](http://www.iqtree.org/).\n",
    "                                    \n",
    "To start, execute the cell below by selecting it and pressing `CTRL+ENTER`and then click in the generated`Run All` button below. Or just go to the **Toolbar** above and select `Run`->`Run all cells`.\n",
    "\n",
    "An **Error** message will be reported after the empty input form in the bottom of the notebook, fill it with your data to proceed. \n",
    "\n",
    ">If in GOOGLE COLAB, `Run All` and `Submit` buttons may not work, in both cases execute cells below through the **Toolbar** menu."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import ipywidgets as widgets\n",
    "#\n",
    "# Run all\n",
    "#\n",
    "def run_all(ev):\n",
    "    display(Javascript('IPython.notebook.execute_cells_below()'))\n",
    "\n",
    "start_running = widgets.Button(description =\"Run All\",\n",
    "                               button_style='success',\n",
    "                               tooltip     ='Click here to start.')\n",
    "start_running.on_click(start_running)\n",
    "\n",
    "display(start_running)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Test if in GOOGLE COLAB enviroment, and if yes, install `ETE3` and `iGraph` packages\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# @title Test if in GOOGLE COLAB enviroment, and if yes, install missing packages\n",
    "try:\n",
    "  import google.colab\n",
    "except ModuleNotFoundError:\n",
    "    in_colab = False\n",
    "else:\n",
    "    in_colab = True\n",
    "    !pip install python-igraph\n",
    "    !pip install ete3\n",
    "    \n",
    "    from google.colab import files\n",
    "    \n",
    "    import requests\n",
    "    functions = requests.get(\n",
    "        'https://raw.githubusercontent.com/lthiberiol/evolSimIndex/master/base_functions.py'\n",
    "    )\n",
    "    with open('base_functions.py', 'w') as functions_handle:\n",
    "        functions_handle.write(functions.text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from io                     import BytesIO, StringIO\n",
    "from IPython.display        import Javascript, FileLink\n",
    "\n",
    "import numpy      as np\n",
    "import pandas     as pd\n",
    "\n",
    "import multiprocessing\n",
    "import itertools\n",
    "import re\n",
    "import base_functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simple parameters and data input interface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Data input buttons\n",
    "evol_dist_source = widgets.Dropdown(\n",
    "    options    =[('',                       0       ), \n",
    "                 ('FASTA files',            'fasta' ), \n",
    "                 ('IQTree \".mldist\" files', 'matrix'), \n",
    "                 ('newick files',           'tree'  )],\n",
    "    disabled   =False,\n",
    "    indent     =False,\n",
    "    value      =0,\n",
    "    layout     ={'width':'auto'}\n",
    ")\n",
    "\n",
    "must_align = widgets.Checkbox(\n",
    "    value   =False,  \n",
    "    disabled=True,\n",
    "    indent  =False,\n",
    "    description='Provided FASTAS are not yet aligned',\n",
    "    layout     ={'width':'auto'}\n",
    ")\n",
    "\n",
    "gene_ids = widgets.Checkbox(\n",
    "    value   =False,  \n",
    "    disabled=False,\n",
    "    indent  =False,\n",
    "    description='Sequences are identified by genome only '\n",
    "    '(all sequences from the same genome have the same name)',\n",
    "    layout     ={'width':'auto'}\n",
    ")\n",
    "\n",
    "min_taxa_overlap = widgets.IntText(value      =5, \n",
    "                                   indent     =False,\n",
    "                                   disabled   =False)\n",
    "\n",
    "genome_gene_sep = widgets.Dropdown(\n",
    "    options    =[('',                                     0  ), \n",
    "                 ('<genome>_<gene>', '_'), \n",
    "                 ('<genome>|<gene>', '|'), \n",
    "                 ('<genome>.<gene>', '.')],\n",
    "    disabled   =False,\n",
    "    indent     =True,\n",
    "    value      =0,\n",
    "    layout     ={'width':'auto'}\n",
    ")\n",
    "\n",
    "input_files = widgets.FileUpload(\n",
    "    accept  ='',   # Accepted file extension e.g. '.txt', '.pdf', 'image/*', 'image/*,.pdf'\n",
    "    multiple=True, # True to accept multiple files upload else False\n",
    "    disabled=True\n",
    ")\n",
    "\n",
    "def toggle_align_widgets(dropdown_source):\n",
    "    input_files.disabled = not dropdown_source.new\n",
    "    \n",
    "    if dropdown_source.new == 'fasta':\n",
    "        must_align.disabled = False\n",
    "#     elif dropdown_source.new == 'example':\n",
    "#         genome_gene_sep.value = '_'\n",
    "#         input_files.disabled  = True\n",
    "    else:\n",
    "        must_align.disabled = True\n",
    "        must_align.value    = False\n",
    "    \n",
    "def toggle_genome_gene_sep(checkbox):\n",
    "    genome_gene_sep.disabled = checkbox.new\n",
    "    if checkbox.new:\n",
    "        genome_gene_sep.value = 0\n",
    "        \n",
    "def clear_uploads(*args):\n",
    "    input_files.value.clear()\n",
    "    input_files._counter = 0\n",
    "    input_files.disabled = False\n",
    "    \n",
    "    evol_dist_source.value = 0\n",
    "    genome_gene_sep.value  = 0\n",
    "    \n",
    "    example.flag = False\n",
    "        \n",
    "clear_button = widgets.Button(description='Clear upload',\n",
    "                              button_style='warning',\n",
    "                              tooltip     ='Click to clear uploaded files')\n",
    "clear_button.on_click(clear_uploads)\n",
    "\n",
    "evol_dist_source.observe(toggle_align_widgets, names='value')\n",
    "gene_ids.observe(toggle_genome_gene_sep,       names='value')\n",
    "\n",
    "\n",
    "#\n",
    "# load example\n",
    "#\n",
    "example = widgets.Button(description='Load example',\n",
    "                         button_style='success',\n",
    "                         tooltip     ='Load example parameters')\n",
    "example.flag = False    \n",
    "def load_example(*args):\n",
    "    evol_dist_source.disabled = False\n",
    "    evol_dist_source.value    = 'matrix'\n",
    "    \n",
    "    genome_gene_sep.value     = '_'\n",
    "    \n",
    "    input_files.disabled      = True\n",
    "    \n",
    "    example.flag = True\n",
    "    \n",
    "example.on_click(load_example)\n",
    "\n",
    "#\n",
    "# submit data\n",
    "#\n",
    "def run_all(ev):\n",
    "    display(Javascript('IPython.notebook.execute_cells_below()'))\n",
    "\n",
    "submit = widgets.Button(description =\"Submit\",\n",
    "                        button_style='success',\n",
    "                        tooltip     ='Click here to continue with provided data')\n",
    "submit.on_click(run_all)\n",
    "\n",
    "#\n",
    "# threads\n",
    "#\n",
    "num_threads = widgets.IntSlider(min=1, \n",
    "                                max=multiprocessing.cpu_count())\n",
    "\n",
    "#\n",
    "# download through colab\n",
    "#\n",
    "if in_colab:\n",
    "    def download(*args):\n",
    "        files.download('Ies.csv')\n",
    "\n",
    "    download_csv = widgets.Button(description='Download Ies.csv',\n",
    "                          button_style='success')\n",
    "    download_csv.on_click(download)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(widgets.HBox([widgets.Label('Source of pairwise distances (there is an \"example\" options): '), \n",
    "                      evol_dist_source]),\n",
    "        must_align,\n",
    "        gene_ids,\n",
    "        widgets.HBox([widgets.Label('Genome and gene ids are separated by which character: '),\n",
    "                      genome_gene_sep]),\n",
    "        \n",
    "        widgets.HBox([widgets.Label('Minimum taxa containing both assessed gene families: '),\n",
    "                      min_taxa_overlap]),\n",
    "        \n",
    "        widgets.HBox([widgets.Label('Number of threads to use: '),\n",
    "                      num_threads]),\n",
    "        \n",
    "#         input_files,\n",
    "        widgets.HBox([input_files, example]),\n",
    "        clear_button,\n",
    "        submit\n",
    "       )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If `Submit` button above is not working, in the **Toolbar** click in `Cell`->`Run All Bellow`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Parsing provided data and parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# @title Breaking \"run all\" if no data was uploadded\n",
    "\n",
    "if not input_files._counter > 1 and not example.flag:\n",
    "    raise ValueError('You must upload at least two files!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if   genome_gene_sep.value == '_':\n",
    "    parse_leaf = re.compile('^(GC[AF]_\\d+(?:\\.\\d)?)[_|](.*)$')\n",
    "elif genome_gene_sep.value == '|':\n",
    "    parse_leaf = re.compile('^(\\S+?)\\|(\\S+)$')\n",
    "elif genome_gene_sep.value == '.':\n",
    "    parse_leaf = re.compile('^(\\d+?)\\.(.*)$')\n",
    "    \n",
    "if min_taxa_overlap.value < 2:\n",
    "    min_taxa_overlap.value = 2\n",
    "\n",
    "corr_evol = base_functions.correlate_evolution(gene_ids       =gene_ids.value,\n",
    "                                               parse_leaf     =parse_leaf,\n",
    "                                               min_taxa_overlap=min_taxa_overlap.value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# @title Loading data...\n",
    "\n",
    "dist_matrices = []\n",
    "group_names   = []\n",
    "\n",
    "if evol_dist_source.value == 'tree':\n",
    "    for file_name, file_itself in input_files.value.items():\n",
    "        dist_matrices.append( \n",
    "           corr_evol.get_matrix_from_tree(file_itself['content'].decode('utf-8')) \n",
    "        )\n",
    "        group_names.append( file_name )\n",
    "        \n",
    "elif evol_dist_source.value == 'matrix':\n",
    "    if not input_files._counter and example.flag:\n",
    "        for file_itself in ['000284', '000302', '000304', '000321', '000528', \n",
    "#                             '000574', '000575', '000595', '000602', '000607',\n",
    "#                             '000611', '000617', '000620', '000621', '000625',\n",
    "                            '000632', '000645', '000647', '000657', '000663']:\n",
    "            dist_matrices.append(corr_evol.load_matrix(\n",
    "                f'https://raw.githubusercontent.com/lthiberiol/evolSimIndex/master/tests/{file_itself}.mldist'\n",
    "            ))\n",
    "            group_names.append( file_itself )\n",
    "    else:\n",
    "        for file_name, file_itself in input_files.value.items():\n",
    "            dist_matrices.append( \n",
    "                corr_evol.load_matrix(BytesIO(file_itself['content'])) \n",
    "            )\n",
    "            group_names.append( file_name )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assessing evolutionary similarity between gene families!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time \n",
    "\n",
    "print(f'Assessing Ies between {len(group_names)} genes')\n",
    "print(f'\\t**using {num_threads.value} threads\\n')\n",
    "\n",
    "matrix_combinations     = itertools.combinations(dist_matrices, 2)\n",
    "group_name_combinations = itertools.combinations(group_names,   2)\n",
    "group_name_combinations = np.array( list(group_name_combinations) )\n",
    "\n",
    "pool    = multiprocessing.Pool(processes=num_threads.value)\n",
    "results = pool.starmap(corr_evol.assess_coevolution, matrix_combinations)\n",
    "pool.close()\n",
    "pool.join()\n",
    "\n",
    "coevol_df = pd.DataFrame(columns=['R_squared', 'Ibc', 'Ies'], \n",
    "                         data   =results)\n",
    "\n",
    "coevol_df['gene1'] = group_name_combinations[:, 0]\n",
    "coevol_df['gene2'] = group_name_combinations[:, 1]\n",
    "\n",
    "coevol_df.to_csv('Ies.csv')\n",
    "\n",
    "if not in_colab:\n",
    "    local_file = FileLink('Ies.csv', result_html_prefix=\"Click here to download: \")\n",
    "    display(local_file)\n",
    "else:\n",
    "    display(download_csv)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
